pkg/controller/etcdproxy/certs.go

```diff
@@ -8,195 +8,297 @@ import (
 
 	"k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	utilerrors "k8s.io/apimachinery/pkg/util/errors"
 
 	etcdstoragev1alpha1 "github.com/xmudrii/etcdproxy-controller/pkg/apis/etcd/v1alpha1"
 	"github.com/xmudrii/etcdproxy-controller/pkg/certs"
+	"k8s.io/apimachinery/pkg/api/errors"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
 )
 
 // ProxyCertificateExpiryAnnotation is annotating are certificates successfully created and stored in the Secret or ConfigMap.
 const ProxyCertificateExpiryAnnotation = "etcd.xmudrii.com/certificate-generated"
 
-// setNewEtcdProxyCertificates generates new self-signed server signer and server certificate/key pair.
-// Self-signed server signer cert is stored in APIServer ConfigMap to be used as CA by etcd, while server cert/key pair
-// is stored in Controller Secret to be used by etcdproxy.
-func (c *EtcdProxyController) setNewEtcdProxyCertificates(etcdstorage *etcdstoragev1alpha1.EtcdStorage) error {
-	currentTime := time.Now
-	r := rand.New(rand.NewSource(currentTime().UnixNano()))
-	serviceUrl := fmt.Sprintf("%s.%s.svc", serviceName(etcdstorage), c.config.ControllerNamespace)
+// ensureClientCertificates ensures there are client CA bundle and client certificates generated and in place.
+// The Client CA bundle is stored in the ConfigMap in the controller namespace.
+// The Client certificates are stored in Secrets defined in the EtcdStorage Spec.
+func (c *EtcdProxyController) ensureClientCertificates(etcdstorage *etcdstoragev1alpha1.EtcdStorage) error {
+	var clientCA *certs.Certificate
+	var errs []error
+	for _, clientCertSecret := range etcdstorage.Spec.ClientCertSecrets {
+		secretClientCert, err := c.kubeclientset.CoreV1().Secrets(clientCertSecret.Namespace).Get(clientCertSecret.Name, metav1.GetOptions{})
+		// TODO: This is hack and we should get rid of it.
+		if errors.IsNotFound(err) {
+			secretClientCert, err = c.kubeclientset.CoreV1().Secrets(clientCertSecret.Namespace).Create(&v1.Secret{ObjectMeta: metav1.ObjectMeta{Name: clientCertSecret.Name, Namespace: clientCertSecret.Namespace}})
+		}
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
 
-	// TODO: implement certificates regeneration. this could require architecture change, because we could require
-	// new secret for storing singer key.
-	serverSigner, err := certs.NewCACertificate(pkix.Name{
-		CommonName: fmt.Sprintf("%s-server-signer-%v", serviceUrl, time.Now().Unix()),
-	}, r.Int63n(100000), currentTime)
-	if err != nil {
-		return err
+		if expiry, ok := secretClientCert.Annotations[ProxyCertificateExpiryAnnotation]; ok {
```

Comment:
you're going to have to keep track of who signed it too.  You can rotate before expiry.  It's ok as a followup.

Reply:
To be decided will this be a follow-up or in this PR. updated the PR description.